lib/events/index.js

```diff
@@ -2036,11 +2175,11 @@ export default function (self) {
       });
     });
 
-    self.dispatchEvent('afterpaste', {
+    self.dispatchEvent('afterfilldown', {
```

Comment:
This feels off to me. Not every paste will be a fill down, will it? Sometimes you're just pasting cells, and sometimes you're just filling cells, and not even pasting the data. But if you emit this event here, you're signalling to the user that every paste you do is _also_ a filldown event, which seems incorrect to me, no?

As an aside, I should note regarding terminology, that "fill down" is only one direction, one can also [fill up, or left, or right](https://support.microsoft.com/en-us/office/fill-a-formula-down-into-adjacent-cells-041edfe2-05bc-40e6-b933-ef48c3f308c6), so to be more correct this event would have to be called `afterfill`:

<img width="517" alt="Fill" src="https://user-images.githubusercontent.com/950979/158096504-3ce0df4a-5adb-4952-803b-3abd55072756.png">

(I should have been a bit sharper about correct naming earlier on, apologies for that!)

But getting back to the main point: I feel like two things are getting mixed up here which to me seem related but also feel distinct: pasting and filling data.

To me, _pasting_ is when we just take whatever is in the copy buffer and overwrite the same number cells in the grid with the pasted data:

![CleanShot 2022-03-14 at 04 02 31](https://user-images.githubusercontent.com/950979/158097654-d47cb6e2-3d7b-45a9-9f42-ccce1869df94.gif)

 _Filling_ is when you need to 'generate' values based on one or more cells in either the currently selected cells or in the copy buffer:

![CleanShot 2022-03-14 at 04 06 04](https://user-images.githubusercontent.com/950979/158098005-74c492a6-1781-4065-9831-c3c8153d5a4d.gif)

I say 'generate' because these values can either all be the same (as above) or based on some inferred pattern, for example repeating or incrementing:

![CleanShot 2022-03-14 at 04 08 27](https://user-images.githubusercontent.com/950979/158098202-3ee2f775-5c53-4089-a67f-0e0459ebf17e.gif)

Notice how the first column is incrementing. The exact logic for generating values is something the user-passed fill function would have to determine how to handle. But in neither of these past two examples am I pasting data, so I would not be expecting an `afterpaste` event.

It gets murkier, though, when you're pasting data into a selection that's larger than the size of your copy buffer:

![CleanShot 2022-03-14 at 04 04 20](https://user-images.githubusercontent.com/950979/158097828-e613b687-c354-4633-9c9e-ec813574a5dd.gif)

In this example, we only have two copied cells in our buffer, but we're pasting into an 11x2 selection. I'm not entirely sure how to view that: is that simply a paste action, a fill action, or a paste-then-fill?

_To me_, intuitively, I think it's the latter: we're pasting two cells (the first row of the selection), and then filling _down_ (in this case) the remaining ten rows. So I'd expect both an `afterpaste` and an `afterfill` event.

If so, then I think the code could be structured a little bit simpler, in that the paste function should do what it has been doing, and only after pasting the full copy buffer needs to determine whether any filling is needed as well (i.e. when the selection is larger than the copy buffer bounds). This way you'd be keeping the filling-logic separate from the pasting-logic.

Would love to hear your thoughts on this!

Reply:
I totally agree. I should have given more thought to the paste data behavior. However, I think we should agree on the terminology used here. 

I believe we should consider pasting a form of filling data because the logic is almost the same. The confusing part might be the name `fillDown`, which we can rename to `fill` to make it more suitable for pasting. 

Also, separating pasting from filling logic might duplicate the work, since, in the case of pasting, the difference between overfilling or just filling is whether we are in selection bounds or not. That is also the reason why `fillDown` function have `source` as an argument. 

If the source is a paste event, and we have enough data to fill in the selection bounds, we consider that a filling operation, and once the data insufficient, we go on to behave like it is an overfilling one. One thing we can avoid doing here is invoking the cell callback when we are just filling. I think it should become less confusing if I don't do that, unless we are overfilling. 

That being said, you have given a great input here, so I will be some making changes like fixing the `afterfilldown` event.

Reply:
Just wanted to get back to you on this: 

> I believe we should consider pasting a form of filling data because the logic is almost the same. The confusing part might be the name fillDown, which we can rename to fill to make it more suitable for pasting.

Agreed!

> Also, separating pasting from filling logic might duplicate the work, since, in the case of pasting, the difference between overfilling or just filling is whether we are in selection bounds or not. That is also the reason why fillDown function have source as an argument.

My intuition tells me there's a larger difference than just that (for instance, we need an if-statement now to check if we've filled any cells to prevent emitting an `afterfill` event when we're pasting), but I'm having a hard time putting this intuition into words — so as long things are testable and debuggable and in the future relatively easy to refactor, which they now seem to be, it's good with me — I'll resolve this conversation!

Reply:
I just saw this. I want to say I no longer think filling should be considered a form pasting because other platforms don't do that either (guessing from the screenshot you sent). Sorry, I should have written this earlier.